AWSAppSyncClient/AppSyncMQTTClient.swift

```diff
@@ -91,59 +99,96 @@ class AppSyncMQTTClient: AWSIoTMQTTClientDelegate {
-        }
-    }
-    
-    
-    /// Returnes updated dictionary
-    /// it removes subscriber from the array
-    ///
-    /// - Parameters:
-    ///   - dictionary: [String: [MQTTSubscritionWatcher]]
-    ///   - subscription: MQTTSubscritionWatcher
-    /// - Returns: [String: [MQTTSubscritionWatcher]]
-    private func updatedDictionary(_ dictionary: [String: [MQTTSubscritionWatcher]] ,
-                                   usingCancelling subscription: MQTTSubscritionWatcher) -> [String: [MQTTSubscritionWatcher]] {
-        
-        return topicSubscribersDictionary.reduce(into: [:]) { (result, element) in
-            result[element.key] = removedSubscriber(array: element.value, of: subscription.getIdentifier())
         }
     }
     
-    
-    
     /// Unsubscribe topic
     ///
     /// - Parameter topic: String
     private func unsubscribeTopic(topic: String) {
         for (client, _)  in mqttClientsWithTopics.filter({ $0.value.contains(topic) }) {
-            client.unsubscribeTopic(topic)
+            switch client.mqttStatus {
+            case .connecting, .connected, .connectionError, .connectionRefused, .protocolError:
+                client.unsubscribeTopic(topic)
+            case .disconnected, .unknown:
+                break
+            }
             mqttClientsWithTopics[client]?.remove(topic)
         }
     }
     
-    /// Removes subscriber from the array using id
-    ///
-    /// - Parameters:
-    ///   - array: [MQTTSubscritionWatcher]
-    ///   - id: Int
-    /// - Returns: updated array [MQTTSubscritionWatcher]
-    private func removedSubscriber(array: [MQTTSubscritionWatcher], of id: Int) -> [MQTTSubscritionWatcher] {
-        return array.filter({$0.getIdentifier() != id })
+    class TopicSubscribers {
+        
+        private var dictionary = [String : NSHashTable<MQTTSubscritionWatcher>]()
+        
+        private var lock = NSLock()
+        
+        subscript(key: String) -> [MQTTSubscritionWatcher]? {
+            get {
+                return synchronized() {
+                    return self.dictionary[key]?.allObjects
+                }
+            }
+        }
+        
+        func add(watcher: MQTTSubscritionWatcher, topics: [String]) {
+            synchronized() {
+                for topic in topics {
+                    if let watchers = self.dictionary[topic] {
+                        watchers.add(watcher)
+                    } else {
+                        let watchers = NSHashTable<MQTTSubscritionWatcher>.weakObjects()
+                        watchers.add(watcher)
+                        self.dictionary[topic] = watchers
+                    }
+                }
+            }
+        }
+        
+        func remove(subscription: MQTTSubscritionWatcher) {
+            synchronized() {
+                dictionary.forEach({ (element) in
+                    element.value.allObjects.filter({ $0.getIdentifier() != subscription.getIdentifier() }).forEach({ (watcher) in
```

Comment:
I think this should be `==` instead of `!=`

```swift
element.value.allObjects.filter({ $0.getIdentifier() == subscription.getIdentifier() }).forEach({ (watcher) in
```